openedx/frontend-app-course-authoring
# [FC-0036] feat: Make tags widget keyboard accessible

## Description

This PR adds the ability to navigate the new "Add Tags" widget using the keyboard, making it fully accessible through the keyboard.

## Supporting Information

Related Tickets:
- Closes https://github.com/openedx/modular-learning/issues/189

## Testing Instructions
1. Run you local dev stack on this branch
1. If you don't already have sample taxonomy/tags data, follow the instructions in this repo to generate sample data: https://github.com/open-craft/taxonomy-sample-data
1. Navigate to a sample course and open the tags drawer by clicking on "Manage Tags"
1. Expand a taxonomy, and navigate to the "Add a Tag" select input using the keyboard by "tabbing" to it
1. Confirm the following:
    1. Once you've navigated to the "Add a Tag" select input the menu should open
    1. Tab again and that should take you to the first element in the menu
    1. Once you're in the tags menu, navigate around using the arrows, confirm that:
        - up/down - go up or down.
        - Left - if leaf or already collapsed: jump to parent tag. Otherwise: collapse the current level.
        - Right - expand the current tag if possible.
        - Enter - if current tag has children: expand/collapse. If leaf: add tag.
        - Space - add tag (or undo adding if it's already added). Same as clicking the checkbox.
    1. When navigating the tags, each tag should be focused as a single widget (the checkbox, the label, and the expand arrow).
    1. When a tag anywhere in the tree is focused, pressing TAB should move the focus to the "Load More" link at the bottom, then to "Cancel" then to "Add Tags". The only way to move among rows/tags in the tag tree area is using the arrow keys. (To test out the "load more", change https://github.com/openedx/frontend-app-course-authoring/blob/6ae9cdac0088117836138e713606c423bec317d5/src/content-tags-drawer/data/api.js#L27 to have a smaller threshold)
    1. Confirm The Load More / Cancel / Add Tags actions must all be usable via keyboard.
    1. Stage a few tags (do not add them yet), then set the focus to the input, by Tabbing back
    1. Confirm that moving the cursor around with the arrow keys will select/focus on different chips. Specifically, the X receives the focus.
    1. Inspect element and check that each staged tag chip's X  has an aria-label
    1. Pressing BACKSPACE in the "Add Tags" text field once will focus on the chip before the cursor. Pressing it a second time will delete that chip. Pressing DELETE will also delete any focused/selected chip.
    1. if you type anything (e.g. types a letter), the focus always jumps to the end (after all the chips). Search keywords cannot be typed before/between the chips.
    1. Inspect the tag elements in the dropdown and confirm that the aria label of the first element in the tag tree dropdown  has: "[Unchecked/implicit/checked] Tag: [Name of the tag]. Use the arrow keys to move among the tags in this taxonomy. Press space to select a tag.", the rest should have "[Unchecked/implicit/checked] Tag: [Name of the tag]."
    1. Confirm that the select menu closes when you Tab out of it, or when you click outside

---
Private-ref: [FAL-3644](https://tasks.opencraft.com/browse/FAL-3644)
